### ISSUE TITLE
Android Auto support

### ISSUE THREAD

**Issue by ahmedre** [quran/quran_android#1086]
Let's consider adding support for Android Auto.

**Comment by MonikaJethani** [comment]
@ahmedre Can I work on this?

**Comment by ahmedre** [comment]
@MonikaJethani please feel free

**Comment by benomaire** [comment]
@MonikaJethani Any updates?

**Comment by farooqu** [comment]
I think it would be nice to consider the desired feature set for Android Auto since it obviously would not be the full app experience.

**Comment by jibraniqbal666** [comment]
I wanna work on this next when i will have some time. @ahmedre :)
- I am thinking of maybe adding ayah by ayah screen, which is recited by desired reciter.
- wanna do it in compose :).

**Comment by ahmedre** [comment]
sure - but before doing so, let's figure out what it will look like and what it will do just so we're aligned.
btw, we are using compose, it's just all hidden in the closed source portion of the repo (the new page rendering is all done in compose today).

**Comment by jibraniqbal666** [comment]
yeah let's do that, how do we proceed for new feature, which needs design.
what do you mean by closed source portion?

**Comment by ahmedre** [comment]
historically, people clone the repo, add some ads on top, and publish as is (keeping our servers, our email contact info, etc). as a result, we get a lot of complaints and so on. to make this a little more difficult, we've started keeping parts of the app in a closed source repo - the open source repo is still the base, but some code isn't there.

the intention longer term is to have a set of libraries that are open sourced that apps can build a Quran app on top of, instead of a full app like is the case today.

with respect to auto - I mean let's think about exactly what we want to do and support (feature wise even not code wise). I imagine something like this would be focused around the audio experience, but curious what you had in mind. I also would think there are probably higher priority issues than this if you're interested can point you to some of them.

**Comment by jibraniqbal666** [comment]
that's understandable.
That's what i was thinking, but we can tackle the higher issues first, point me to it. :)
<img width="928" alt="image" src="https://user-images.githubusercontent.com/30576574/163892533-f4651de3-4cd9-4546-a78f-16b26a66a82e.png">

**Comment by dasyad00** [comment]
Hi, I'm also interested in helping out with this PR.

Is the vision mainly to have media controls? Do we want the user to only start the Quran playback from the phone or also the car infotainment? I feel the latter will require a lot of UI/UX considerations especially to reduce distractions for the driver.

**Comment by benomaire** [comment]
@jibraniqbal666 Showing text might be tricky, because it is something that changes frequently, and some ayat are long and might require scrolling text? The following [quote](https://developers.google.com/cars/design/design-foundations/interaction-principles) is relevant:

> Avoid displaying dynamic visual information unrelated to driving, such as videos and auto-scrolling text.

@dasyad00 

> Is the vision mainly to have media controls?

I think yes. Or at least, it could be a good start.

> Do we want the user to only start the Quran playback from the phone or also the car infotainment?

also the car infotainment IMHO.

I would imagine buttons 
- to quickly start audio playback from the last position, 
- to start playback from a random Surah and Ayah

and also two lists for choosing the Surah and Ayah (similar to the following screenshot from quran.com)
![screenshot-](https://user-images.githubusercontent.com/69304392/163988082-b552a075-f0b9-4eba-a394-2350fe1c72cf.png)

**Comment by ahmedre** [comment]
in my head, an Android Auto experience will only be around audio, since the person driving should not be using the display while driving, and can't really read ayat on it easily - I wouldn't expect a reading experience on it?

**Comment by jibraniqbal666** [comment]
it's audio, we are just showing ayah as another visual, but i think driver might get distracted from it. then we could just see Spotify auto which might just have play pause skip and title of the surah. @ahmedre

**Comment by ahmedre** [comment]
would be good like you mentioned to see what other apps do on Android Auto to get some ideas

**Comment by jibraniqbal666** [comment]
<img width="237" alt="image" src="https://user-images.githubusercontent.com/30576574/164454351-257e6ea9-23f1-4358-81a2-f9891ab997c9.png">

**Comment by jibraniqbal666** [comment]
@ahmedre ⬆️

**Comment by nacer80** [comment]
السلام عليكم عليكم عليكم وبركاته 
I'm using Bluetooth option to listen to Quran while driving and it's very useful 
.

![IMG_20220423_195744.jpg](https://user-images.githubusercontent.com/60948528/164942653-1c550e3b-b1a9-4759-97d9-b879cabb2539.jpg)

**Comment by ahmedre** [comment]
Sure something similar to what either you @jibraniqbal666 or @nacer80 shared works - we should have all this info in the notification object we use for playback

**Comment by nacer80** [comment]
@ahmedre the last screenshot was from my screen car connected with quran_android app by bluetooth.
Amazing job, thanks to all contributers.
-still if you add tashkeel to Surah "سورة" ..."سُورَةُ"

**Comment by jibraniqbal666** [comment]
let's prototype some ui in compose :) @ahmedre, if we don't have compose in public repo, should i add it?
btw i am planning to add 2 screens
- List to select the Surah.
- Play the Surah.

**Comment by jibraniqbal666** [comment]
> @ahmedre the last screenshot was from my screen car connected with quran_android app by bluetooth. Amazing job, thanks to all contributers. -still if you add tashkeel to Surah "سورة" ..."سُورَةُ"

you should create a separate issue for tashkeel as i won't be focusing on bluetooth one for now :).

**Comment by ahmedre** [comment]
yes please feel free - versions are there in build.gradle so let's use those since that's what the rest of the project is using -

```
    composeVersion = '1.1.0-rc01'
    composeCompilerVersion = '1.1.0-rc02'
```

**Comment by jibraniqbal666** [comment]
@ahmedre we need api 23 to support Android Auto :)

**Comment by ahmedre** [comment]
should be able to just enable auto features on 23+ without bumping minSdk

**Comment by jibraniqbal666** [comment]
@ahmedre unfortunately no auto emulator is working in M1 mac :(, i have to find another way.

**Comment by ahmedre** [comment]
https://twitter.com/alsutton/status/1510994278553690116?s=21&t=8gTC8MkfrhlRMepF7X5Osg

**Comment by ahmedre** [comment]
There was something but the link is gone now 😔 https://issuetracker.google.com/issues/219764014

**Comment by jibraniqbal666** [comment]
@dasyad00 can you help us on atleast, just the structure. :), if you have intel mac?

**Comment by jibraniqbal666** [comment]
Unfortunately, there is no Android Auto emulator on M1 yet, hence i can't even run and see the app, i would leave this issue stale for now and focus on something else.
@ahmedre

**Comment by YousufSohail** [comment]
@jibraniqbal666 @ahmedre I have intel mac. I can help. Could you please let me know in what branch we have work related to android auto which I can run on an emulator?

**Comment by jibraniqbal666** [comment]
@YousufSohail https://github.com/jibraniqbal666/quran_android/tree/feature/base-android-auto it's just base, you need to run it.
- we could do this task collaboratively, as i can build stuff, but can't test it.
- first we merge base with basic ui (you can add) as i can't test them.
- then i add feature on your base, which you can test. 
:)

**Comment by NadimMussaDaud** [comment]
Assalam aleikum, i would like to have some update about the implementation of this feature so that I can help. Djazakallah khair.

**Comment by FardinEajdani25** [comment]
Assalamu Alaykum. Can I work on this PR?

**Comment by ahmedre** [comment]
@jibraniqbal666's work in progress is in #2648. will let him reply to this issue if he needs help.

**Comment by ahmedre** [comment]
This is now merged. Some things that are left:

1. generate album art per qari - just a colored box with the initials of the qari, since otherwise, the grid view is very dull.
2. maintain a list of the last 3 qaris and the last playing file
3. update to support resume playback method after point 2 above is complete
4. update categories to include all Qaris after 1 & 2 are done
5. update search to also search sura names instead of just qari names

**Comment by Arakmmis** [comment]
@ahmedre I would like to tackle the 5 things you mentioned if that's ok. I'll start with what seems to me to be the easiest to get out of the way: the first two points. Would you like me to create issues for each or keep the discussion in this thread?

**Comment by ahmedre** [comment]
@Arakmmis masha'Allah sounds awesome! Yes let's make new issues for those and we can link them back to this one.

**Comment by ahmedre** [comment]
another issue someone reported which is 100% true and on point - there's no "syncing" between the phone and the car in the sense that if you play something on the phone and get in the car, the Android Auto app doesn't show what's playing and allow you to control it. it's completely separate at the moment. unfortunately this will likely be a larger change (likely needing to merge the two approaches of handling audio at some point).

**Comment by AhmedGamal92** [comment]
SA @ahmedre and everyone 👋 

Currently working on a solution to sync the app audio service with android auto.

The idea is to have the android auto playback using the same downloaded content as if you play from the main app. Now the question I have where I would need your input on is what happens when the user selects a sura that's not downloaded yet (and not having the always stream option) ? In the main app we trigger a download first and given with android auto users would mostly be on data connection we can't really ask whether we download or not as while interacting with the app.

The options I have in my mind so far:
1. The app auto switch to stream mode when a non downloaded surah is selected, and we add a setting for it so the user can disable if needed and if disabled nothing is played (probably with showing some error). This can be combined with indicating with an icon whether a surah is downloaded or not for a better ux.
2. Follow a similar idea as 1 but to auto download instead with a setting as well. Not sure about this one as that can result in some unintentional high data volume consumption. 
3. Add a setting for both 1&2 to give the user complete control, probably defaulting to streaming and then the user can change that setting in the app to either don't play or download. 
4. Do nothing from android auto and play only downloaded surah, with showing a message or so to the user to do that through the phone for surah not downloaded. But I'd say that's quite a frustrating option and the worst ux.


A follow up question that's also a bit related is should we auto play next surahs or we stick with the selected sura only and stop after ?

**Comment by ahmedre** [comment]
wa3laikum alsalam,
jazakumAllah khairan - today, we're just using streaming in this case, and I'd vote to continue doing so for now. the only change we should ideally make in this regard is to check if the file is downloaded, and, if so, use the local one, otherwise use the streaming version. I wouldn't bother making this a setting tbh.

for the follow-up question, auto play 👍

**Comment by AhmedGamal92** [comment]
@ahmedre Fair! Agreed, I'll continue with that then.

**Comment by hazemelraffiee** [comment]
Assalamu Alaykum @ahmedre,

I'd like to work on the following from your list:

1. **Generate album art per qari** — colored box with the qari's initials
2. **Maintain a list of the last 3 qaris and the last playing file**

Were these already picked up by @Arakmmis, or are they still available? Happy to create separate issues for each and link them back here.

**Comment by ahmedre** [comment]
wa3laikum alsalam wa ra7matullah,
1 is already done and shipped in production.

you are welcome to pick up 2 though.
walsalam 3alaikum

**Comment by hazemelraffiee** [comment]
Opened #3567 — adds a "Recently Played" tab to the Android Auto browse tree.

It tracks the last 5 played sura+qari combinations and shows them alongside the Qaris tab. The list updates in real time as new tracks play. Tested on a Pixel 6 with the Desktop Head Unit (USB transport mode).

This covers item 2 from @ahmedre's list (maintain a list of the last played qaris/suras). Item 1 (album art per qari) was already shipped.
